packages/asset-swapper/src/utils/market_operation_utils/fill_optimizer.ts

```diff
@@ -32,24 +32,27 @@ export class FillsOptimizer {
     public optimize(fills: Fill[], input: BigNumber, upperBoundPath?: Fill[]): Fill[] | undefined {
         this._currentRunCount = 0;
         this._optimalPath = upperBoundPath;
-        this._optimalPathOutput = upperBoundPath ? getPathOutput(upperBoundPath, input) : ZERO_AMOUNT;
+        this._optimalPathAdjustedOutput = upperBoundPath ? getPathAdjustedOutput(upperBoundPath, input) : ZERO_AMOUNT;
         const ctx = {
             currentPath: [],
             currentPathInput: ZERO_AMOUNT,
-            currentPathOutput: ZERO_AMOUNT,
+            currentPathAdjustedOutput: ZERO_AMOUNT,
             currentPathFlags: 0,
         };
         // Visit all valid combinations of fills to find the optimal path.
         this._walk(fills, input, ctx);
-        return this._optimalPath;
+        if (this._optimalPath) {
+            return sortFillsByAdjustedRate(this._optimalPath, this._shouldMinimize);
+        }
+        return undefined;
```

Comment:
Curiously, there was no strong guarantee that the optimizer would return fills in descending exchange rate. This was fine because later, after creating the orders, we would sort them by rate there. Sorting the orders afterwards won't take into account tx + gas costs though, so they now need to be done at the fill level.